README.md
docker-machine-driver-xhyve
===========================

| **Travis CI**                           | **Releases**                          |
|:---------------------------------------:|:-------------------------------------:|
| [![Travis CI][travis-badge]][travis-ci] | [![Releases][release-badge]][release] |

[![GA][ga-badge]][ga]

libmachine driver plugin for [xhyve](https://github.com/mist64/xhyve) native OS X Hypervisor

Master branch inherited from [nathanleclaire/docker-machine-xhyve](https://github.com/nathanleclaire/docker-machine-xhyve). Thanks [@nathanleclaire](https://github.com/nathanleclaire) :)  
If you have issues or pull-requests, Desired to be posted to this repository.

Screencast
----------

[![asciinema](imgs/launch.png)](https://asciinema.org/a/29930)

Requirements
------------

### docker-machine, minikube or minishift

-	https://github.com/docker/machine
-	https://github.com/kubernetes/minikube
-	https://github.com/minishift/minishift

docker-machine-driver-xhyve using libmachine plugin model.

**Please do not post the issue of this repository to docker/machine, kubernetes/minikube and minishift/minishift**  
It will interfere with the development of docker-machine, minikube or minishift.  
If you were doubt problem either, please post to this repository [issues](https://github.com/zchee/docker-machine-driver-xhyve/issues).

-	docker-machine

	-	See https://github.com/docker/machine/releases

-	minikube

	-	See https://github.com/kubernetes/minikube/releases

-	minishift

	-	See https://github.com/minishift/minishift/releases

Install
-------

Use [Homebrew/brew](https://github.com/Homebrew/brew):

```sh
$ brew install docker-machine-driver-xhyve

# docker-machine-driver-xhyve need root owner and uid
$ sudo chown root:wheel $(brew --prefix)/opt/docker-machine-driver-xhyve/bin/docker-machine-driver-xhyve
$ sudo chmod u+s $(brew --prefix)/opt/docker-machine-driver-xhyve/bin/docker-machine-driver-xhyve
```

Use `go` with `make`:  
If you want to support qcow2 disk image format, need install [mirage/ocaml-qcow](https://github.com/mirage/ocaml-qcow). See [docker/hyperkit#building](https://github.com/docker/hyperkit#building).

```sh
# Need Go 1.5 vendoring support
$ export GO15VENDOREXPERIMENT=1

$ go get -u -d github.com/zchee/docker-machine-driver-xhyve
$ cd $GOPATH/src/github.com/zchee/docker-machine-driver-xhyve

# Install qcow-format for qcow2 disk image format
$ brew install opam libev
$ opam init
$ eval `opam config env`
$ opam install uri qcow-format io-page.1.6.1 conf-libev

# Install docker-machine-driver-xhyve binary into /usr/local/bin
$ make install

# docker-machine-driver-xhyve need root owner and uid
$ sudo chown root:wheel /usr/local/bin/docker-machine-driver-xhyve
$ sudo chmod u+s /usr/local/bin/docker-machine-driver-xhyve
```

We use [Glide](https://github.com/Masterminds/glide) for dependency management.

```sh
$ go get github.com/Masterminds/glide
```

This will install the glide binary into `$GOPATH/bin`.

Updating Dependencies

If your work requires a change to the dependencies, you need to update the Glide
configuration.

- Edit glide.yaml to change the dependencies as needed.
- Delete `glide.lock` and re-create the vendor directory by running `make vendor`. Glide will recognize that there is no lock file and recalculatethe required dependencies.
- Check-in the updated `glide.yaml` and `glide.lock` files.
- Test that everything still compiles with the new lock file in place by running make clean && make.

Note: In some cases the Glide cache located under ~/.glide/cache can get corrupted. If you seeing Glide errors during `make vendor`, you can clear the Glide cache via `glide cc`.

Usage
-----

### Available flags

| Flag name                        | Environment variable           | Type   | Default                                                                                                                              |
|----------------------------------|--------------------------------|--------|--------------------------------------------------------------------------------------------------------------------------------------|
| `--xhyve-boot2docker-url`        | `XHYVE_BOOT2DOCKER_URL`        | string | `$HOME/.docker/machine/cache/boot2docker.iso`                                                                                        |
| `--xhyve-cpu-count`              | `XHYVE_CPU_COUNT`              | int    | `1`                                                                                                                                  |
| `--xhyve-memory-size`            | `XHYVE_MEMORY_SIZE`            | int    | `1024`                                                                                                                               |
| `--xhyve-disk-size`              | `XHYVE_DISK_SIZE`              | int    | `20000`                                                                                                                              |
| `--xhyve-uuid`                   | `XHYVE_UUID`                   | int    | `''`                                                                                                                                 |
| `--xhyve-boot-cmd`               | `XHYVE_BOOT_CMD`               | string | See [AUTOMATED_SCRIPT.md](https://github.com/boot2docker/boot2docker/blob/master/doc/AUTOMATED_SCRIPT.md#extracting-boot-parameters) |
| `--xhyve-boot-kernel`            | `XHYVE_BOOT_KERNEL`            | string | `''`                                                                                                                                 |
| `--xhyve-boot-initrd`            | `XHYVE_BOOT_INITRD`            | string | `''`                                                                                                                                 |
| `--xhyve-qcow2`                  | `XHYVE_QCOW2`                  | bool   | `false`                                                                                                                              |
| `--xhyve-virtio-9p`              | `XHYVE_VIRTIO_9P`              | bool   | `false`                                                                                                                              |
| `--xhyve-experimental-nfs-share` | `XHYVE_EXPERIMENTAL_NFS_SHARE` | bool   | `false`                                                                                                                              |

#### `--xhyve-boot2docker-url`

The URL(Path) of the boot2docker image.  
By default, use cached iso file path.

#### `--xhyve-cpu-count`

Number of CPUs to use the create the VM.  
If set `-1`, use logical CPUs usable by the current process.

#### `--xhyve-memory-size`

Size of memory for the guest.

#### `--xhyve-disk-size`

Size of disk for the guest (MB).

#### `--xhyve-uuid`

The UUID for the machine.  
By default, generate and use ramdom UUID. See [xhyve/uuid.go](https://github.com/zchee/docker-machine-driver-xhyve/blob/master/xhyve/uuid.go)

#### `--xhyve-boot-cmd`

Booting xhyve kexec commands.  
By default, use  
`loglevel=3 user=docker console=ttyS0 console=tty0 noembed nomodeset norestore waitusb=10 base host=boot2docker`

#### `--xhyve-boot-kernel`

Booting kernel file path.  
By default, will automatically parses the file path using `(vmlinu[xz]|bzImage)[\d]*`.

#### `--xhyve-boot-initrd`

Booting initrd file path.  
By default, will automatically parses the `initrd` contains file path.

#### `--xhyve-qcow2`

Use `qcow2` disk format.  
If you using minikube, `CONFIG_VIRTIO_BLK=y` support is included in minikube-iso as of version v0.0.6.

#### `--xhyve-rawdisk`

Use a simple 'raw disk' format and virtio-blk driver for storage.
This may be significantly faster for I/O intensive applications, at the potential cost of data durability.

#### `--xhyve-virtio-9p`

Enable `virtio-9p` folder share.  
If you using docker-machine, `CONFIG_NET_9P=y` support is included in boot2docker as of version v1.10.2.

#### `--xhyve-experimental-nfs-share`

Enable `NFS` folder sharing.

Known isuue
-----------

### Does not clean up the vmnet when remove a VM

Currently, `docker-machine-driver-xhyve` does not clean up the `dhcpd_leases`.  
like,

```sh
# Running xhyve vm. for example, assign 192.168.64.1
$ docker-machine create --driver xhyve xhyve-test
        |
# Send ACPI signal(poweroff) signal over the ssh
$ docker-machine rm xhyve-test
        |
# Re-create xhyve vm, will assign 192.168.64.2
docker-machine create --driver xhyve xhyve-test
```

It will assigned to 192.168.64.**2**. If create another vm, assigned to 192.168.64.**3**.  
But 192.168.64.**1** are not using anyone.

`vmnet.framework` seems to have decide the IP based on below files

-	`/var/db/dhcpd_leases`
-	`/Library/Preferences/SystemConfiguration/com.apple.vmnet.plist`

So, If you want to reset IP database, please remove it manually. but **very risky**.  
Note that `vmnet.framework` shared net address range is `192.168.64.1` ~ `192.168.64.255`. You can make 255 vm.

I will implement the clean-up process after understanding the `vmnet.framework`.

### Can't launch on macOS 10.11.4 build 15E27e

Mac OS X 10.11.4 build `15E27e` has a **Hypervisor.framework bug**.  
This is Apple's bug.  
But, Apple has been fixed build `15E33e`.

If you launch the `docker-machine-driver-xhyve` on build 15E27e, will displayed

```sh
open : no such file or directory
```

and, In original `xhyve`,

```sh
hv_vm_create failed
```

See

-	https://asciinema.org/a/34798 (15E27e)
-	https://asciinema.org/a/34797 (15E33e)

Could you report?
-----------------

I'm very anxious whether other users(except me) are able to launch the xhyve.  
So, if you were able to launch the xhyve use docker-machine-driver-xhyve, Could you post a report to this issue thread?  
https://github.com/zchee/docker-machine-driver-xhyve/issues/18

If macOS launched by the `Vagrant`, can be build, but will not be able to launch the Hypervisor.  
The cause probably because backend vm (Virtualbox, VMWare, parallels...) to hide the CPU infomation.

In the case of VMWare,

```sh
$ system_profiler SPHardwareDataType

CHANGELOG.md
Change Log
==========

[v0.3.3](https://github.com/zchee/docker-machine-driver-xhyve/tree/v0.3.3) (2017-06-29)
=======================================================================================

[Full Changelog](https://github.com/zchee/docker-machine-driver-xhyve/compare/v0.3.2...v0.3.3)

Auto generated Change Log
-------------------------

**Closed issues:**

- GetState returns running after host reboot [\#179](https://github.com/zchee/docker-machine-driver-xhyve/issues/179)
- MAC addresses ending in element with a leading 0 are not mapped to an IP [\#162](https://github.com/zchee/docker-machine-driver-xhyve/issues/162)

**Merged pull requests:**

- xhyve: bump version 0.3.3 [\#184](https://github.com/zchee/docker-machine-driver-xhyve/pull/184) ([zchee](https://github.com/zchee))
- ci/travis: fix opam install packages to locked io-page to 1.6.1 [\#182](https://github.com/zchee/docker-machine-driver-xhyve/pull/182) ([zchee](https://github.com/zchee))
- Add Minishift as user [\#181](https://github.com/zchee/docker-machine-driver-xhyve/pull/181) ([gbraad](https://github.com/gbraad))
- Fix \#179 add process name check to getState [\#180](https://github.com/zchee/docker-machine-driver-xhyve/pull/180) ([gbraad](https://github.com/gbraad))
- Ignore setupmounts [\#174](https://github.com/zchee/docker-machine-driver-xhyve/pull/174) ([r2d4](https://github.com/r2d4))

[v0.3.2](https://github.com/zchee/docker-machine-driver-xhyve/tree/v0.3.2) (2017-04-21)
=======================================================================================

[Full Changelog](https://github.com/zchee/docker-machine-driver-xhyve/compare/v0.3.1...v0.3.2)

Auto generated Change Log
-------------------------

**Closed issues:**

-	Cannot build with Go and Make [\#167](https://github.com/zchee/docker-machine-driver-xhyve/issues/167)
-	Project Is dead? [\#166](https://github.com/zchee/docker-machine-driver-xhyve/issues/166)
-	\[Question\] Why use `go generate` for libhyperkit? [\#165](https://github.com/zchee/docker-machine-driver-xhyve/issues/165)
-	fatal error: 'stdlib.h' file not found [\#160](https://github.com/zchee/docker-machine-driver-xhyve/issues/160)
-	Build broken when homebrew is installed in a non-default location [\#157](https://github.com/zchee/docker-machine-driver-xhyve/issues/157)

**Merged pull requests:**

-	Update readme to document rawdisk [\#173](https://github.com/zchee/docker-machine-driver-xhyve/pull/173) ([dalehamel](https://github.com/dalehamel)\)
-	xhyve: bump version 0.3.2 [\#172](https://github.com/zchee/docker-machine-driver-xhyve/pull/172) ([zchee](https://github.com/zchee)\)
-	Don't hardcode the docker user id [\#171](https://github.com/zchee/docker-machine-driver-xhyve/pull/171) ([dalehamel](https://github.com/dalehamel)\)
-	ci/circle: remove circle.yml for now [\#170](https://github.com/zchee/docker-machine-driver-xhyve/pull/170) ([zchee](https://github.com/zchee)\)
-	Add support for using a raw disk for storage [\#169](https://github.com/zchee/docker-machine-driver-xhyve/pull/169) ([dalehamel](https://github.com/dalehamel)\)
-	Fixed MAC cleaning regular expression [\#163](https://github.com/zchee/docker-machine-driver-xhyve/pull/163) ([juamedgod](https://github.com/juamedgod)\)
-	Wait for IP before setting up mounts [\#161](https://github.com/zchee/docker-machine-driver-xhyve/pull/161) ([r2d4](https://github.com/r2d4)\)
-	ci/circle: update circle.yml to macOS based build environment [\#159](https://github.com/zchee/docker-machine-driver-xhyve/pull/159) ([zchee](https://github.com/zchee)\)
-	Move the NFS and 9p mounting code from Create to Start. [\#158](https://github.com/zchee/docker-machine-driver-xhyve/pull/158) ([dlorenc](https://github.com/dlorenc)\)

[v0.3.1](https://github.com/zchee/docker-machine-driver-xhyve/tree/v0.3.1) (2016-12-05)
=======================================================================================

[Full Changelog](https://github.com/zchee/docker-machine-driver-xhyve/compare/v0.3.0...v0.3.1)

Auto generated Change Log
-------------------------

**Implemented enhancements:**

-	alternative boot2docker ISO image \(minikube-iso\) support [\#135](https://github.com/zchee/docker-machine-driver-xhyve/issues/135)

**Fixed bugs:**

-	alternative boot2docker ISO image \\(minikube-iso\\) support [\#135](https://github.com/zchee/docker-machine-driver-xhyve/issues/135)

**Closed issues:**

-	Extract kernel options from the syslinux.conf file if exist. [\#150](https://github.com/zchee/docker-machine-driver-xhyve/issues/150)
-	Recompile on Go 1.7.1 to work on macOS Sierra \(10.12\) [\#137](https://github.com/zchee/docker-machine-driver-xhyve/issues/137)

**Merged pull requests:**

-	xhyve: Bump v0.3.1 [\#154](https://github.com/zchee/docker-machine-driver-xhyve/pull/154) ([zchee](https://github.com/zchee)\)
-	xhyve: Fix included '\tappend\s' in kexec boot option [\#153](https://github.com/zchee/docker-machine-driver-xhyve/pull/153) ([zchee](https://github.com/zchee)\)

xhyve.go
package xhyve

import (
	"archive/tar"
	"bytes"
	"errors"
	"fmt"
	"io/ioutil"
	"net"
	"os"
	"os/exec"
	"regexp"
	"runtime"
	"strconv"
	"strings"
	"syscall"
	"time"

	"github.com/docker/machine/libmachine/drivers"
	"github.com/docker/machine/libmachine/log"
	"github.com/docker/machine/libmachine/mcnflag"
	"github.com/docker/machine/libmachine/mcnutils"
	"github.com/docker/machine/libmachine/ssh"
	"github.com/docker/machine/libmachine/state"
	"github.com/johanneswuerbach/nfsexports"
	"github.com/zchee/docker-machine-driver-xhyve/vmnet"
)

const (
	isoFilename           = "boot2docker.iso"
	isoMountPath          = "b2d-image"
	defaultBoot2DockerURL = ""
	defaultBootCmd        = "loglevel=3 user=docker console=ttyS0 console=tty0 noembed nomodeset norestore waitusb=10 base host=boot2docker"
	defaultCPU            = 1
	defaultCaCertPath     = ""
	defaultDiskSize       = 20000
	defaultMacAddr        = ""
	defaultMemory         = 1024
	defaultPrivateKeyPath = ""
	defaultUUID           = ""
	defaultNFSShare       = false
)

type Driver struct {
	*drivers.BaseDriver
	Boot2DockerURL string
	BootCmd        string
	CPU            int
	CaCertPath     string
	DiskSize       int64
	MacAddr        string
	Memory         int
	PrivateKeyPath string
	UUID           string
	NFSShare       bool
}

var (
	ErrMachineExist    = errors.New("machine already exists")
	ErrMachineNotExist = errors.New("machine does not exist")
)

// NewDriver creates a new VirtualBox driver with default settings.
func NewDriver(hostName, storePath string) *Driver {
	return &Driver{
		BaseDriver: &drivers.BaseDriver{
			MachineName: hostName,
			StorePath:   storePath,
		},
		Boot2DockerURL: defaultBoot2DockerURL,
		BootCmd:        defaultBootCmd,
		CPU:            defaultCPU,
		CaCertPath:     defaultCaCertPath,
		DiskSize:       defaultDiskSize,
		MacAddr:        defaultMacAddr,
		Memory:         defaultMemory,
		PrivateKeyPath: defaultPrivateKeyPath,
		UUID:           defaultUUID,
		NFSShare:       defaultNFSShare,
	}
}

// RegisterCreateFlags registers the flags this driver adds to
// "docker hosts create"
func (d *Driver) GetCreateFlags() []mcnflag.Flag {
	return []mcnflag.Flag{
		mcnflag.StringFlag{
			EnvVar: "XHYVE_BOOT2DOCKER_URL",
			Name:   "xhyve-boot2docker-url",
			Usage:  "The URL of the boot2docker image. Defaults to the latest available version",
			Value:  "",
		},
		mcnflag.IntFlag{
			EnvVar: "XHYVE_CPU_COUNT",
			Name:   "xhyve-cpu-count",
			Usage:  "Number of CPUs for the machine (-1 to use the number of CPUs available)",
			Value:  defaultCPU,
		},
		mcnflag.IntFlag{
			EnvVar: "XHYVE_MEMORY_SIZE",
			Name:   "xhyve-memory-size",
			Usage:  "Size of memory for host in MB",
			Value:  defaultMemory,
		},
		mcnflag.IntFlag{
			EnvVar: "XHYVE_DISK_SIZE",
			Name:   "xhyve-disk-size",
			Usage:  "Size of disk for host in MB",
			Value:  defaultDiskSize,
		},
		mcnflag.StringFlag{
			EnvVar: "XHYVE_BOOT_CMD",
			Name:   "xhyve-boot-cmd",
			Usage:  "Command of booting kexec protocol",
			Value:  defaultBootCmd,
		},
		mcnflag.BoolFlag{
			EnvVar: "XHYVE_EXPERIMENTAL_NFS_SHARE",
			Name:   "xhyve-experimental-nfs-share",
			Usage:  "Setup NFS shared folder (requires root)",
		},
	}
}

func (d *Driver) GetMachineName() string {
	return d.MachineName
}

func (d *Driver) GetSSHHostname() (string, error) {
	return d.GetIP()
}

func (d *Driver) GetSSHKeyPath() string {
	return d.ResolveStorePath("id_rsa")
}

func (d *Driver) GetSSHPort() (int, error) {
	if d.SSHPort == 0 {
		d.SSHPort = 22
	}

	return d.SSHPort, nil
}

func (d *Driver) GetSSHUsername() string {
	if d.SSHUser == "" {
		d.SSHUser = "docker"
	}

	return d.SSHUser
}

func (d *Driver) DriverName() string {
	return "xhyve"
}

func (d *Driver) SetConfigFromFlags(flags drivers.DriverOptions) error {
	d.Boot2DockerURL = flags.String("xhyve-boot2docker-url")
	d.CPU = flags.Int("xhyve-cpu-count")
	d.Memory = flags.Int("xhyve-memory-size")
	d.DiskSize = int64(flags.Int("xhyve-disk-size"))
	d.BootCmd = flags.String("xhyve-boot-cmd")
	d.SwarmMaster = flags.Bool("swarm-master")
	d.SwarmHost = flags.String("swarm-host")
	d.SwarmDiscovery = flags.String("swarm-discovery")
	d.SSHUser = "docker"
	d.SSHPort = 22
	d.NFSShare = flags.Bool("xhyve-experimental-nfs-share")

	return nil
}

func (d *Driver) GetURL() (string, error) {
	ip, err := d.GetIP()
	if err != nil {
		return "", err
	}
	if ip == "" {
		return "", nil
	}
	return fmt.Sprintf("tcp://%s:2376", ip), nil
}

func (d *Driver) GetIP() (string, error) {
	if d.IPAddress != "" {
		return d.IPAddress, nil
	}

	return d.getIPfromDHCPLease()
}

func (d *Driver) GetState() (state.State, error) {
	log.Infof("Getting to VM state...")
	pid, err := d.GetPid()
	if err != nil {
		// TODO: If err instead of nil, will be occurred error when first GetState() of Start()
		return state.Error, nil
	}

	proc, err := os.FindProcess(int(pid))
	if err != nil {
		return state.Error, err
	}

	if err := proc.Signal(syscall.Signal(0)); err != nil {
		return state.Stopped, nil
	}

	return state.Running, nil
}

// Print driver version, Check VirtualBox version
func (d *Driver) PreCreateCheck() error {
	//TODO:libmachine PLEASE output driver version API!
	v := Version
	c := GitCommit
	log.Debugf("===== Docker Machine %s Driver Version %s (%s) =====\n", d.DriverName(), v, c)

	ver, err := vboxVersionDetect()
	if ver == "" && err == nil {
		return nil
	}
	if err != nil {
		return fmt.Errorf("Error detecting VBox version: %s", err)
	}
	if !strings.HasPrefix(ver, "5") {
		return fmt.Errorf("Virtual Box version 4 or lower will cause a kernel panic" +
			"if xhyve tries to run. You are running version: " +
			ver +
			"\n\t Please upgrade to version 5 at https://www.virtualbox.org/wiki/Downloads")
	}
	return nil
}

func (d *Driver) Create() error {
	b2dutils := mcnutils.NewB2dUtils(d.StorePath)
	if err := b2dutils.CopyIsoToMachineDir(d.Boot2DockerURL, d.MachineName); err != nil {
		return err
	}

	log.Infof("Creating SSH key...")
	if err := ssh.GenerateSSHKey(d.GetSSHKeyPath()); err != nil {
		return err
	}

	log.Infof("Creating VM...")
	if err := os.MkdirAll(d.ResolveStorePath("."), 0755); err != nil {
		return err
	}

	log.Infof("Extracting vmlinuz64 and initrd.img from %s...", isoFilename)
	if err := d.extractKernelImages(); err != nil {
		return err
	}

	log.Infof("Generating %dMB disk image...", d.DiskSize)
	if err := d.generateDiskImage(d.DiskSize); err != nil {
		return err
	}

	// Fix file permission root to current user for vmnet.framework
	log.Infof("Fix file permission...")
	os.Chown(d.ResolveStorePath("."), syscall.Getuid(), syscall.Getegid())
	files, _ := ioutil.ReadDir(d.ResolveStorePath("."))
	for _, f := range files {
		log.Debugf(d.ResolveStorePath(f.Name()))
		os.Chown(d.ResolveStorePath(f.Name()), syscall.Getuid(), syscall.Getegid())
	}

	log.Infof("Generate UUID...")
	d.UUID = uuidgen()
	log.Debugf("Generated UUID: %s", d.UUID)

	log.Infof("Convert UUID to MAC address...")
	rawUUID, err := d.getMACAdress()
	if err != nil {
		return err
	}
	d.MacAddr = trimMacAddress(rawUUID)
	log.Debugf("Converted MAC address: %s", d.MacAddr)

	log.Infof("Starting %s...", d.MachineName)
	if err := d.Start(); err != nil {
		return err
	}
	log.Infof("Waiting for VM to come online...")

	var ip string
	for i := 1; i <= 60; i++ {
		ip, err = d.getIPfromDHCPLease()
		if err != nil {
			log.Debugf("Not there yet %d/%d, error: %s", i, 60, err)
			time.Sleep(2 * time.Second)
			continue
		}

		if ip != "" {
			log.Debugf("Got an ip: %s", ip)
			break
		}
	}

	if ip == "" {
		return fmt.Errorf("Machine didn't return an IP after 120 seconds, aborting")
	}

	// We got an IP, let's copy ssh keys over
	d.IPAddress = ip

	// Setup NFS sharing
	if d.NFSShare {
		err = d.setupNFSShare()
		if err != nil {
			log.Errorf("NFS setup failed: %s", err.Error())
		}
	}

	return nil
}

func (d *Driver) Start() error {
	pid := d.ResolveStorePath(d.MachineName + ".pid")
	if _, err := os.Stat(pid); err == nil {
		os.Remove(pid)
	}

	args := d.xhyveArgs()
	args = append(args, "-F", fmt.Sprintf("%s", pid))

	log.Debug(args)

	cmd := exec.Command(os.Args[0], args...)
	cmd.Stdout = &bytes.Buffer{}
	cmd.Stderr = &bytes.Buffer{}

	err := cmd.Start()
	if err != nil {
		return err
	}

	go func() {
		err := cmd.Wait()
		if err != nil {
			log.Error(err, cmd.Stdout, cmd.Stderr)
		}
	}()

	return nil
}

func (d *Driver) Stop() error {
	log.Infof("Stopping %s ...", d.MachineName)
	if err := d.SendSignal(syscall.SIGTERM); err != nil {
		return err
	}

	for {
		s, err := d.GetState()
		if err != nil {
			return err
		}
		if s == state.Running {
			time.Sleep(1 * time.Second)
		} else {
			break
		}
	}

	d.IPAddress = ""

	return nil
}

func (d *Driver) Remove() error {
	s, err := d.GetState()
	if err != nil {
		if err == ErrMachineNotExist {
			log.Infof("machine does not exist, assuming it has been removed already")
			return nil
		}
		return err
	}
	if s == state.Running {
		if err := d.Stop(); err != nil {
			return err
		}

		if _, err := nfsexports.Remove("", d.nfsExportIdentifier()); err != nil {
			log.Errorf("failed removing nfs share: %s", err.Error())
		}

		if err := nfsexports.ReloadDaemon(); err != nil {
			log.Errorf("failed reload nfs daemon: %s", err.Error())
		}
	}
	return nil
}

func (d *Driver) Restart() error {
	s, err := d.GetState()
	if err != nil {
		return err
	}

	if s == state.Running {
		if err := d.Stop(); err != nil {
			return err
		}
	}

	if err := d.Start(); err != nil {
		return err
	}

	return nil
}

func (d *Driver) Kill() error {
	log.Infof("Killing %s ...", d.MachineName)
	if err := d.SendSignal(syscall.SIGKILL); err != nil {
		return err
	}

	return nil
}

func (d *Driver) setMachineNameIfNotSet() {
	if d.MachineName == "" {
		d.MachineName = fmt.Sprintf("docker-machine-unknown")
	}
}

func (d *Driver) getIPfromDHCPLease() (string, error) {
	currentip, err := vmnet.GetIPAddressByMACAddress(d.MacAddr)
	log.Debugf(currentip)

	if currentip == "" {
		return "", fmt.Errorf("IP not found for MAC %s in DHCP leases", d.MacAddr)
	}

	log.Debugf("IP found in DHCP lease table: %s", currentip)
	return currentip, err
}

func (d *Driver) publicSSHKeyPath() string {
	return d.GetSSHKeyPath() + ".pub"
}

func (d *Driver) extractKernelImages() error {
	log.Debugf("Mounting %s", isoFilename)

	err := hdiutil("attach", d.ResolveStorePath(isoFilename), "-mountpoint", d.ResolveStorePath("b2d-image"))
	if err != nil {
		return err
	}

	volumeRootDir := d.ResolveStorePath(isoMountPath)
	vmlinuz64 := volumeRootDir + "/boot/vmlinuz64"
	initrd := volumeRootDir + "/boot/initrd.img"

	log.Debugf("Extracting vmlinuz64 into %s", d.ResolveStorePath("."))
	if err := mcnutils.CopyFile(vmlinuz64, d.ResolveStorePath("vmlinuz64")); err != nil {
		return err
	}
	log.Debugf("Extracting initrd.img into %s", d.ResolveStorePath("."))
	if err := mcnutils.CopyFile(initrd, d.ResolveStorePath("initrd.img")); err != nil {
		return err
	}
	log.Debugf("Unmounting %s", isoFilename)
	if err := hdiutil("detach", volumeRootDir); err != nil {
		return err
	}

	return nil
}

func (d *Driver) generateDiskImage(count int64) error {
	output := d.ResolveStorePath(d.MachineName)

	if err := hdiutil("create", "-megabytes", fmt.Sprintf("%d", d.DiskSize), output); err != nil {
		return err
	}

	tarBuf, err := d.generateKeyBundle()
	if err != nil {
		return err
	}

	file, err := os.OpenFile(output+".dmg", os.O_WRONLY, 0644)
	if err != nil {
		return err
	}
	defer file.Close()
	file.Seek(0, os.SEEK_SET)
	_, err = file.Write(tarBuf.Bytes())
	if err != nil {
		return err
	}
	file.Close()

	return nil
}

// Make a boot2docker userdata.tar key bundle
func (d *Driver) generateKeyBundle() (*bytes.Buffer, error) {
	magicString := "boot2docker, please format-me"

	buf := new(bytes.Buffer)
	tw := tar.NewWriter(buf)

	// magicString first so the automount script knows to format the disk
	file := &tar.Header{Name: magicString, Size: int64(len(magicString))}
	if err := tw.WriteHeader(file); err != nil {
		return nil, err
	}
	if _, err := tw.Write([]byte(magicString)); err != nil {
		return nil, err
	}
	// .ssh/key.pub => authorized_keys
	file = &tar.Header{Name: ".ssh", Typeflag: tar.TypeDir, Mode: 0700}
	if err := tw.WriteHeader(file); err != nil {
		return nil, err
	}
	pubKey, err := ioutil.ReadFile(d.publicSSHKeyPath())
	if err != nil {
		return nil, err
	}
	file = &tar.Header{Name: ".ssh/authorized_keys", Size: int64(len(pubKey)), Mode: 0644}
	if err := tw.WriteHeader(file); err != nil {
		return nil, err
	}
	if _, err := tw.Write([]byte(pubKey)); err != nil {
		return nil, err
	}
	file = &tar.Header{Name: ".ssh/authorized_keys2", Size: int64(len(pubKey)), Mode: 0644}
	if err := tw.WriteHeader(file); err != nil {
		return nil, err
	}
	if _, err := tw.Write([]byte(pubKey)); err != nil {
		return nil, err
	}
	if err := tw.Close(); err != nil {
		return nil, err
	}
	return buf, nil
}

// Setup NFS share
func (d *Driver) setupNFSShare() error {
	nfsConfig := fmt.Sprintf("/Users %s -alldirs -maproot=root", d.IPAddress)

	if _, err := nfsexports.Add("", d.nfsExportIdentifier(), nfsConfig); err != nil {
		return err
	}

	if err := nfsexports.ReloadDaemon(); err != nil {
		return err
	}

	hostIP, err := vmnet.GetNetAddr()
	if err != nil {
		return err
	}

	bootScriptName := "/var/lib/boot2docker/bootlocal.sh"
	bootScript := fmt.Sprintf("#/bin/bash\\n"+
		"sudo mkdir -p /Users\\n"+
		"sudo /usr/local/etc/init.d/nfs-client start\\n"+
		"sudo mount -t nfs -o noacl,async %s:/Users /Users\\n", hostIP)

	writeScriptCmd := fmt.Sprintf("echo -e \"%s\" | sudo tee %s && sudo chmod +x %s && %s",
		bootScript, bootScriptName, bootScriptName, bootScriptName)

	if _, err := drivers.RunSSHCommandFromDriver(d, writeScriptCmd); err != nil {
		return err
	}

	return nil
}

func (d *Driver) nfsExportIdentifier() string {
	return fmt.Sprintf("docker-machine-driver-xhyve %s", d.MachineName)
}

func (d *Driver) GetPid() (int, error) {
	p, err := ioutil.ReadFile(d.ResolveStorePath(d.MachineName + ".pid"))
	if err != nil {
		return 0, err
	}

	pid, err := strconv.ParseInt(string(p), 10, 64)
	if err != nil {
		return 0, err
	}

	return int(pid), nil
}

func (d *Driver) SendSignal(sig os.Signal) error {
	pid, err := d.GetPid()
	if err != nil {
		return err
	}

	proc, err := os.FindProcess(int(pid))
	if err != nil {
		return err
	}

	if err := proc.Signal(sig); err != nil {
		return err
	}

	return nil
}

//Trimming "0" of the ten's digit
func trimMacAddress(rawUUID string) string {
	re := regexp.MustCompile(`[0]([A-Fa-f0-9][:])`)
	mac := re.ReplaceAllString(rawUUID, "$1")

	return mac
}

func (d *Driver) xhyveArgs() []string {
	uuid := d.UUID
	vmlinuz := d.ResolveStorePath("vmlinuz64")
	initrd := d.ResolveStorePath("initrd.img")
	iso := d.ResolveStorePath(isoFilename)
	img := d.ResolveStorePath(d.MachineName + ".dmg")
	bootcmd := d.BootCmd

	cpus := d.CPU
	if cpus < 1 {
		cpus = int(runtime.NumCPU())
	}

	return []string{
		"xhyve",
		"-A",
		"-U", fmt.Sprintf("%s", uuid),
		"-c", fmt.Sprintf("%d", cpus),
		"-m", fmt.Sprintf("%dM", d.Memory),
		"-l", "com1",
		"-s", "0:0,hostbridge",
		"-s", "31,lpc",
		"-s", "2:0,virtio-net",
		"-s", fmt.Sprintf("3,ahci-cd,%s", iso),
		"-s", fmt.Sprintf("4,virtio-blk,%s", img),
		"-f", fmt.Sprintf("kexec,%s,%s,%s", vmlinuz, initrd, bootcmd)}
}

func (d *Driver) getMACAdress() (string, error) {
	args := append(d.xhyveArgs(), "-M")

	stdout := bytes.Buffer{}

	cmd := exec.Command(os.Args[0], args...) // TODO: Should be possible without exec
	cmd.Stdout = &stdout
	if err := cmd.Run(); err != nil {
		return "", err
	}

	mac := bytes.TrimPrefix(stdout.Bytes(), []byte("MAC: "))
	mac = bytes.TrimSpace(mac)

	hw, err := net.ParseMAC(string(mac))
	if err != nil {
		return "", err
	}
	return hw.String(), nil
}
